lib/TPP/Transforms/LowerPacksAndUnpacksWithoutTranspose.cpp

```diff
@@ -0,0 +1,185 @@
+    maps[operandIdx] = AffineMapAttr::get(normalizedIndexingMap);
+    genericOp.setIndexingMapsAttr(ArrayAttr::get(rewriter.getContext(), maps));
+  }
+}
+
+struct LowerPackOnInputsFoldingTranspose
+    : public OpRewritePattern<linalg::GenericOp> {
+  // Is only called with single-user packOp operands, so callback can always
+  // find the (use by the) linalg.generic that is the target of the pattern.
+  using ControlFn = std::function<bool(tensor::PackOp)>;
+  ControlFn controlFn;
+
+  LowerPackOnInputsFoldingTranspose(MLIRContext *context,
+                                    ControlFn controlFn = nullptr,
+                                    PatternBenefit benefit = 1)
+      : OpRewritePattern(context, benefit), controlFn(std::move(controlFn)) {}
+
+  LogicalResult matchAndRewrite(linalg::GenericOp genericOp,
+                                PatternRewriter &rewriter) const override {
+    bool modifiedAnOperand = false;
+    for (auto &&[operandIdx, inOperand] :
+         llvm::enumerate(genericOp.getInputs())) {
+      auto packOp =
+          dyn_cast_if_present<tensor::PackOp>(inOperand.getDefiningOp());
+
+      if (!packOp || !packOp->hasOneUse() || (controlFn && !controlFn(packOp)))
+        continue;
+
+      lowerPackAndFoldTranspose(packOp, genericOp, operandIdx, rewriter);
+
+      modifiedAnOperand = true;
+    }
+
+    return modifiedAnOperand ? success() : failure();
+  }
+};
+
+struct LowerPackUnpackOnOutputFoldingTranspose
+    : public OpRewritePattern<linalg::GenericOp> {
+  // Is only called with single-user packOp operands, so callback can always
+  // find the (use by the) linalg.generic that is the target of the pattern.
+  using ControlFn = std::function<bool(tensor::PackOp, tensor::UnPackOp)>;
+  ControlFn controlFn;
+
+  LowerPackUnpackOnOutputFoldingTranspose(MLIRContext *context,
+                                          ControlFn controlFn = nullptr,
+                                          PatternBenefit benefit = 1)
+      : OpRewritePattern(context, benefit), controlFn(std::move(controlFn)) {}
+
+  LogicalResult matchAndRewrite(linalg::GenericOp genericOp,
+                                PatternRewriter &rewriter) const override {
+    bool modifiedAnOperand = false;
+    size_t numInputs = genericOp.getInputs().size();
+    for (auto &&[outOperandIdx, outOperand] :
+         llvm::enumerate(genericOp.getOutputs())) {
+      size_t operandIdx = numInputs + outOperandIdx;
+      auto result = genericOp->getResult(outOperandIdx);
+
+      if (!result.hasOneUse())
+        continue;
+
+      auto packOp =
+          dyn_cast_if_present<tensor::PackOp>(outOperand.getDefiningOp());
+      auto unpackOp =
+          llvm::dyn_cast<tensor::UnPackOp>(*(result.getUsers().begin()));
+
+      if (!packOp || !packOp->hasOneUse() || !unpackOp ||
+          unpackOp.getInnerDimsPos() != packOp.getInnerDimsPos() ||
+          unpackOp.getOuterDimsPerm() != packOp.getOuterDimsPerm() ||
```

Comment:
nit: in case of identity `getOuterDimsPerm` can be empty so it's not completely reliable

Reply:
Fixed.

Reply:
A sanity check on this would be appreciated - thanks!

Reply:
Approach is fine, just a small fix in the logic is needed.